compiler/rustc_builtin_macros/src/contracts.rs

```diff
@@ -0,0 +1,92 @@
+use rustc_ast::token;
+use rustc_ast::tokenstream::{DelimSpacing, DelimSpan, Spacing, TokenStream, TokenTree};
+use rustc_errors::ErrorGuaranteed;
+use rustc_expand::base::{AttrProcMacro, ExtCtxt};
+use rustc_span::symbol::{sym, Symbol};
+use rustc_span::Span;
+
+pub struct ExpandRequires;
+
+impl AttrProcMacro for ExpandRequires {
+    fn expand<'cx>(
+        &self,
+        ecx: &'cx mut ExtCtxt<'_>,
+        span: Span,
+        annotation: TokenStream,
+        annotated: TokenStream,
+    ) -> Result<TokenStream, ErrorGuaranteed> {
+        expand_requires_tts(ecx, span, annotation, annotated)
+    }
+}
+
+fn expand_requires_tts(
+    _ecx: &mut ExtCtxt<'_>,
+    attr_span: Span,
+    annotation: TokenStream,
+    annotated: TokenStream,
+) -> Result<TokenStream, ErrorGuaranteed> {
+    let mut new_tts = Vec::with_capacity(annotated.len());
+    let mut cursor = annotated.into_trees();
+
+    // NOTE: this will break if you have a `{ ... }` syntactically prior to the fn body,
+    // e.g. if a `const { ... }` can occur in a where clause. We need to do something
+    // smarter for injecting this code into the right place.
```

Comment:
using `register_attr` will resolve this, as you get access to an https://doc.rust-lang.org/nightly/nightly-rustc/rustc_expand/base/enum.Annotatable.html directly, so you can easily find the body and create wrappers for it and such. It's how we handle tests and benchmarks

Reply:
I think I've tried this myself too, but doesn't `register_attr` expect either a function or something that has a `expand` signature of

```rs
fn expand(
    &self,
    ecx: &mut ExtCtxt<'_>,
    span: Span,
    meta_item: &ast::MetaItem,
    item: Annotatable,
)
```

Specifically, `ast::MetaItem` the annotation here limits what syntax is permitted within a hypothetical `#[contracts::requires(<requires_expr>)]`? It would fix the `{ ... }` breaking issue, but then wouldn't it introduce a new `<requires_expr>` can no longer be a complex expression issue?

Reply:
hmm... true... but if we want to permit expressions there we should teach `NestedItem` about it anyway (gated appropriately that it won't be permitted without feature gates even behind `#[cfg(FALSE)]`)

Reply:
Hmm.

So... rather than trying to build this atop `Attr` (which is oriented around TokenTree to TokenTree transformations), it sounds like you are encouraging that I try to extend the functionality of the expanders we feed into `LegacyAttr` (namely functions that implement the `MultiItemModifier` trait, such as the signature @jieyouxu wrote above, which will then involve expanding the definition of `ast::MetaItem` (or one of the types it uses, such as `NestedMetaItem`) to now be able to take as inputs ... token trees? asts?

Do I have that right?

Reply:
Yea, we already support `#[foo = bar]` where `bar` is an arbitrary expression, we just error on it in ast lowering. Seems fine to teach that to parse `#[foo(someexpr)]`, too. But I'd want a @petrochenkov  vibe check on that before touching it